docs/developers_notes/basis_module.md

```diff
@@ -0,0 +1,130 @@
+# The Basis Module
+
+The `neurostatslib.basis` module provides objects that allow users to construct and evaluate basis functions of various types. The classes are hierarchically organized as follows:
+
+```
+Abstract Class Basis
+|
+├─ Abstract Subclass AdditiveBasis
+│
+├─ Abstract Subclass MultiplicativeBasis
+│
+├─ Abstract Subclass SplineBasis
+│   │
+│   └─ Subclass MSplineBasis
+│
+├─ Abstract Subclass RaisedCosineBasis
+│   │
+│   ├─ Subclass RaisedCosineBasisLinear
+│   │
+│   └─ Subclass RaisedCosineBasisLog
+│
+└─ Class OrthExponentialBasis
+```
+
+We've used abstract classes to ensure that any basis object inheriting from the superclass will implement the abstract methods of the superclass. This guarantees that if the inputs and outputs of those methods conform to the requirements specified by the abstract superclass, any new basis class implementation will integrate seamlessly with the basis objects already implemented in the library.
+
+The user only needs to instantiate the non-abstract subclasses located at the bottom of the hierarchy. These classes provide two public methods, `evaluate` and `evaluate_on_grid`, both of which are defined in the superclass `Basis`. These methods perform checks on both the input provided by the user and the output of the evaluation to ensure correctness, and are thus considered "safe."
+
+Additionally, a user can combine multiple objects that are subclasses of Basis, creating a higher-dimensional basis set object. Basis objects can be combined using the `__add__` and `__mul__` methods, which allow for the following syntax:
+
+```
+basis_add = basis_a + basis_b # calls basis_a.__add__(basis_b)
+
+basis_add = basis_a * basis_b # calls basis_b.__mul_(basis_b)
+```
+
+## The Class `neurostatslib.basis.Basis`
+
+### The Public Method `evaluate`
+
+The `evaluate` method checks input consistency and evaluates the basis function at some sample points. It accepts one or more numpy arrays as input, which represent the sample points at which the basis will be evaluated, and performs the following steps:
+
+1. Checks that the inputs all have the same sample size `N`, and raises a `ValueError` if this is not the case.
+2. Checks that the number of inputs matches what the basis being evaluated expects (e.g., one input for a 1-D basis, N inputs for an N-D basis, or the sum of N 1-D bases), and raises a `ValueError` if this is not the case.
+3. Checks that the output of the method does not exceed `gb_limit` GB of memory size. This size limit is specified at class initialization.
```

Comment:
it is approximate the size of an array of size N as  (N x size of the array elements).
for example, for the size of an array of 1000 numpy.int64, you would get 8000 bytes instead if you generate the array and you call sys.getsizeof(array) you would get 8112 bytes on my machine
(112 bytes is the size of the instantiated empty array that you would get with sys.getsizeof(np.array([],dtype=np.int64))).

The key is the approximation is accurate enough given that the overlay of the numpy & jax array is very small

Reply:
Gotcha, that makes sense (and I think is worth describing somewhere, in docs or docstring).

But in general, like I think I've mentioned, I'm uncomfortable with this 3rd step. I don't think I'm aware of another scientific python package that implements this check? I.e., you can easily create a torch tensor, numpy array, or xarray that's too big for memory, and I'm assuming there's a good reason for that. My guess is that this step could cause it to break in weird ways, and it's not really doing what the user wants --- they'd want to make sure *their whole analysis pipeline* (or at least the GLM portion that's written in our package) doesn't exceed memory (rather than just the basis matrix), but there's no way we can guarantee that. 

Ultimately, it's up to you, but I think we should take the fact that this doesn't exist in other scientific packages, despite being easy to approximate and relevant, as a strong warning sign (happy to be proven wrong if this is somewhere!)